src/unittest/python/execution_tests.py

```diff
@@ -346,6 +346,49 @@ def test_ensure_after_action_is_executed_when_task_is_executed(self):
         verify(action).execute({})
         verify(task).execute(any(), {})
 
+    def test_ensure_after_action_run_always_is_executed_when_task_fails(self):
+        task = mock(name="task", dependencies=[])
+        when(task).execute(any(), {}).thenRaise(ValueError("simulated task error"))
+        action = mock(name="action", execute_before=[], execute_after=["task"], run_always=True)
+
+        self.execution_manager.register_action(action)
+        self.execution_manager.register_task(task)
+        self.execution_manager.resolve_dependencies()
+
+        try:
+            self.execution_manager.execute_task(task)
+        except Exception as e:
+            self.assertTrue(type(e), ValueError)
+            self.assertEquals(str(e), "simulated task error")
+
+        verify(action).execute({})
+        verify(task).execute(any(), {})
+
+    def test_ensure_after_action_run_always_is_executed_when_action_fails(self):
+        task = mock(name="task", dependencies=[])
+        action_regular = mock(name="action_regular", execute_before=[], execute_after=["task"], run_always=False)
+        when(action_regular).execute({}).thenRaise(ValueError("simulated action error"))
+        action_cleanup = mock(name="action_cleanup", execute_before=[], execute_after=["task"], run_always=True)
+        action_after_cleanup = mock(name="action_after_cleanup", execute_before=[], execute_after=["task"],
+                                    run_always=False)
+
+        self.execution_manager.register_action(action_regular)
+        self.execution_manager.register_action(action_cleanup)
+        self.execution_manager.register_action(action_after_cleanup)
+        self.execution_manager.register_task(task)
+        self.execution_manager.resolve_dependencies()
+
+        try:
+            self.execution_manager.execute_task(task)
+        except Exception as e:
+            self.assertTrue(type(e), ValueError)
```

Comment:
This is wrong, I think... it will always coerce the type of e to `True` and takes the type `ValueError` as an error message if the first value was not true (but this never happens).
Also the test will not fail if the exception is not raised so we need an `else` block.